stanford/roles/cluster/tasks/emr.yml

```diff
@@ -0,0 +1,23 @@
+---
+- name: Lookup/create emr slave security group
+  ec2_group:
+    # Tags added in Ansible 2.4
+    name: "{{ CLUSTER_SECURITY_GROUP }}-slave"
+    description: "{{ CLUSTER_SECURITY_GROUP_DESCRIPTION }}-slave"
+    vpc_id: "{{ DEPLOYMENT_VPC_ID }}"
+    region: "{{ CLUSTER_REGION }}"
+    rules: 
+    - proto: icmp
+      from_port: 0
+      to_port: 65535
+      cidr_ip: "{{ DEPLOYMENT_CIDR }}"
+    - proto: tcp
+      from_port: 0
+      to_port: 65535
+      cidr_ip: "{{ DEPLOYMENT_CIDR }}"
+    - proto: udp
+      from_port: 0
+      to_port: 65535
+      cidr_ip: "{{ DEPLOYMENT_CIDR }}"
```

Comment:
My reading of this [1] leads me to believe we can clamp down on the slave and master security groups a bit:

It looks like both slave and master need to open every port on ICMP, TCP, and UDP _just_ to other slaves and masters.
So I think we can lock this down like:

```yaml
rules:
- proto: icmp
  from_port: 0
  to_port: 65535
  cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.205.0/24"
- proto: icmp
  from_port: 0
  to_port: 65535
  cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.206.0/24"
- proto: tcp
  from_port: 0
  to_port: 65535
  cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.205.0/24"
- proto: tcp
  from_port: 0
  to_port: 65535
  cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.206.0/24"
- proto: udp
  from_port: 0
  to_port: 65535
  cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.205.0/24"
- proto: udp
  from_port: 0
  to_port: 65535
  cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.206.0/24"
```

and as you're already doing above, the master also needs 8443 to specific AWS IPs.

It looks like we have to double up each of the protocols (there are two stanzas each for ICMP, TCP, and UDP), in order to specify multiple sources (each of the two subnets), for now, until we upgrade Ansible to v2.4:

> Rule sources list support was added in version 2.4. This allows to define multiple sources per source type as well as multiple source types per rule. Prior to 2.4 an individual source is allowed. In version 2.5 support for rule descriptions was added.

[1] https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-man-sec-groups.html

Reply:
Since most of the rules are re-used between the two groups, I think we can move it to a variable which is then shared between master and slave (with master having its 8443 rule added too) like:


```yaml
# somewhere shared, like `roles/cluster/defaults/main.yml`
cluster_emr_internal_group_rules:
  - proto: icmp
    from_port: 0
    to_port: 65535
    cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.205.0/24"
  - proto: icmp
    from_port: 0
    to_port: 65535
    cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.206.0/24"
  - proto: tcp
    from_port: 0
    to_port: 65535
    cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.205.0/24"
  - proto: tcp
    from_port: 0
    to_port: 65535
    cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.206.0/24"
  - proto: udp
    from_port: 0
    to_port: 65535
    cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.205.0/24"
  - proto: udp
    from_port: 0
    to_port: 65535
    cidr_ip: "{{ DEPLOYMENT_CIDR_BASE }}.206.0/24"
cluster_emr_internal_group_rules_slave: "{{ cluster_emr_internal_group_rules }}"
cluster_emr_internal_group_rules_master:
  <<: *cluster_emr_internal_group_rules
  - proto: udp
    from_port: 8443
    to_port: 8443
    cidr_ip: WHATEVER_THE_FIRST_AWS_IP_IS
  # then repeat for the other AWS IPs..
```

And then you can just use these later like:

```yaml
CLUSTER_SECURITY_GROUP_RULES:
  # ...
  emr: "{{ cluster_emr_internal_group_rules_master }}"
```

and

```yaml
ec2_group:
  rules: "{{ cluster_emr_internal_group_rules_slave }}"
```

You can see an example of YAML variable combining here:
 - https://github.com/Stanford-Online/configuration/blob/master/playbooks/roles/edxapp/defaults/main.yml#L1100-L1105

Reply:
And actually, I guess instead of continually rewriting the IPs like `"{{ DEPLOYMENT_CIDR_BASE }}.206.0/24"`, you could instead do this, to pull it dynamically.

`"{{ CLUSTER_SECURITY_GROUP_RULES['emr'][0]['cidr']"`
and
`"{{ CLUSTER_SECURITY_GROUP_RULES['emr'][1]['cidr']"`

Reply:
Yes, I saw that doc also.

As for the AWS ip's. I didn't find them in any doc. I lifted them out of the SGs that have been previously created for CME and data-vpc Insights.

So, there is a no-code option for doing all of this. Using the UI, launch an EMR cluster and the SG's will be created automatically. Then terminated the cluster but re-use the SG's for all other clusters.